Stack space reservation

paulidale opened the pull request:
- [x] documentation is added or updated
- [x] tests are added or updated

Add a call to allow space in a _stack_ to be reserved ahead of time.

richsalz commented on ssl/ssl_cert.c:
```diff
@@ -468,7 +469,8 @@ STACK_OF(X509_NAME) *SSL_dup_CA_list(const STACK_OF(X509_NAME) *sk)
+    sk_X509_NAME_reserve(ret, num);
+    for (i = 0; i < num; i++) {
         name = X509_NAME_dup(sk_X509_NAME_value(sk, i));
         if (name == NULL || !sk_X509_NAME_push(ret, name)) {
```
If reserver guarantees that push can't fail for memory, why are we checking?  Is it this kind of thing that makes it a wip?

paulidale replied:
The code either needs to check the return from the reserve call or from each individual insertion.  It doesn't need to do both.  I took the simpler (lazier?) route by not checking the reserve return code.  If the reserve fails, it is still possible the insertions might succeed.

dot-asm replied:
I'm actually with Rich on this one. In sense it would be more appropriate to check for return value from reserve call, rather than attempt to check individual insertions. Even though there is a chance that individual assertions can work (because another thread would free some memory after reservation attempt?), it's more likely to fail midway.

paulidale commented:
I've addressed the same allocations as #4379 here.

paulidale commented:
Travis error is unrelated.

levitte commented on include/openssl/stack.h:
```diff
@@ -10,6 +10,8 @@
 #ifndef HEADER_STACK_H
 # define HEADER_STACK_H
 
+# include <stddef.h>
```
Why is this needed?

paulidale replied:
I started allowing an expansion with a size_t argument but decided that was insane because the count is an int.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -118,32 +118,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
+        || (size_t)n > SIZE_MAX / sizeof(char *) - st->num)
+        return 0;
+
+    target = st->num + n * sizeof(char *);
```
Why n is multiplied by `sizeof(char*)`?

paulidale replied:
Gone now.  Left over from the size_t based expansion's earlier code.
Although it was probably wrong then too :(

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -118,32 +118,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
+    size_t target;
+    size_t num_alloc = st->num_alloc;
```
This is redundant assignment. It asks for redundant question why. Well, initially it was just question, and then it turned redundant. But I wish it wasn't posed to start with. I mean remove redundant assignment.

paulidale replied:
It's gone now.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -118,32 +118,57 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
 
         /* Avoid overflow due to multiplication by sizeof(char *) */
-        if (doub_num_alloc > SIZE_MAX / sizeof(char *))
+        if (sz > SIZE_MAX / sizeof(char *))
```
Actually this is not the only boundary. It makes no sense allocate more than INT_MAX[+1] elements, because maximum index in int.

On related note one can wonder if doubling is actually right strategy. Well, it gets kind of "schizophrenic" here. Doubling is probably right in start, then it should rather switch to linear, but you are probably unlikely to even need that many entries. And since amount of entries is not large, then checks for SIZE_MAX feels excessive...

paulidale replied:
By my reading of the standard, it looks like SIZE_MAX can be less than INT_MAX.
I've limited the size to INT_MAX (pushed separately but will be squashed).

As for the doubling, I wondered about this too.  I don't know the right strategy.  Do we have data as to the size stacks can grow to inside OpenSSL?  It is a public API, so we can't know for sure.

paulidale replied:
As for the size increase, the best reference I've found so far is _Why Are Vectors Efficient?_ by Andrew Koenig from the _Journal of Object-Orientated Programming_ volume 11, number 5, published in 1998.  I've not been able to locate the actual article but the result is presented in [this comp.land.c++moderated article](https://groups.google.com/d/msg/comp.lang.c++.moderated/asH_VojWKJw/Lo51JEmLVboJ).

The conclusion is that growth factors greater than:
![Φ = 1.618...](https://latex.codecogs.com/gif.latex?%5CPhi&space;=&space;%5Cfrac{1&space;&plus;&space;sqrt(5)}{2}&space;%5Capprox&space;1.618)
aren't great.  I've found other discussions that suggest linear growth is never great, so I'm going to try exponential growth with a factor of 1.5.

dot-asm replied:
> By my reading of the standard, it looks like SIZE_MAX can be less than INT_MAX.

But in all practical cases it's vice versa and what I'm talking about is rather relation between `SIZE_MAX/sizeof(void *)` vs. `INT_MAX` than `SIZE_MAX` and `INT_MAX`.

> I've limited the size to INT_MAX (pushed separately but will be squashed).

Formally speaking the limit should be `(size_t)INT_MAX+1`. Indeed, last permitted index being INT_MAX means that there are INT_MAX+1 elements allocated. Here is another relevant question. Imagine going back to doubling for a moment, and imagine you wanted to reserve INT_MAX-13 elements. With INT_MAX as cap it would fail, because algorithm would search for closest 2^n larger than requested amount but it will fail check for INT_MAX. For wrong reason. Well, one case of course argue that INT_MAX-13 should never happen and we might as well let it slide. But it does make code counter-intuitive. Check for INT_MAX+1 would address this. But now consider proposed 1.5 growth factor... I don't want to calculate problematic values, what I'm rather trying to ask is if `return 0` as outcome of this test is right? Maybe setting setting  to INT_MAX+1 is more adequate? And by "right" and "adequate" I rather mean rather "least counter-intuitive" than "practical"...

dot-asm replied:
It seems that I'm wrong about INT_MAX+1 thing. stack_st.num is amount of elements, not last index, and it's integer. This means that last possible index is INT_MAX-1, not INT_MAX, as I suggested. But the question about whether or not corner cases similar to INT_MAX-13 should be treated as valid (again, not from practical viewpoint, but formal) remains...

paulidale replied:
I think I've addressed the INT_MAX-13 concern, I limit growth to INT_MAX now.

I left the check against SIZE_MAX/sizeof(char *).  There is an earlier check for n+st->num against INT_MAX that prevents that being too large and the loop is checking for INT_MAX.  These prevent num_alloc being greater than INT_MAX but it is still possible that size_t cannot handle that many pointers (e.g. when sizeof(int) == sizeof(size_t)).  Thus, I think the SIZE_MAX check is correct.

paulidale commented:
Growth by 1.5 done.

dot-asm commented:
As for 1.5. Do we recognize that `x + x/2` is more exact expression than `(x/2)*3`? I mean do we choose latter even if we recognize former?

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -12,6 +12,34 @@
+/*
+ * The initial number of nodes in the array.
+ */
+static const int min_nodes = 4;
```
I was meaning to ask why do you prefer static const over pre-processor macros? I'm not saying it's wrong/inappropriate, just curious over any particular reason. I can imagine it being handy in debugger, because you can examine the value [as opposite to remembering it or consulting likely another file]...

paulidale replied:
Yes to debugging.  It also makes the compiler type check (which can be done in a macro but often isn't).  Compilers have optimised these away for ages now so there is no penalty either for performance or space.

I also think it's a bit neater this way.  I generally find expressions a little easier to read when they're mostly in lower case.

So no compelling reason for or against really.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -12,6 +12,34 @@
+ * to unity could be problematic for some sizes (e.g. 6/5 with a size of 18).
+ */
+static const int growth_numerator = 3;
+static const int growth_denominator = 2;
```
Question here is all [supported] compilers manage to optimize division as shift...

dot-asm replied:
I suppose this also answers the remark about `x+x/2`. I mean as long as one wants to keep the options open, one has to accept that `(x/2)*3` is less exact. Well, it's matter of taste, since assertion is that *any* factor less than 1.618 would do...

richsalz replied:
Maybe have a define or function to get the next size, rather than two contants.

paulidale replied:
I was aware that the division by two became a shift, the expansion shouldn't be a common occurrence so the cost of a division would be acceptable -- some compilers will change this to a multiplication (_Division by invariant integers using multiplication_ by Granlund and Montgomery).

I also realised that `(x/2)*3` is low by _one_ for odd `x`.  Correcting this will add complexity to the overflow checking and that's something I was trying to avoid -- it is difficult enough to reason about these checks and an error in them has consequences.  The latter is the main reason I avoided the 8/5 option which is arguably slightly better (being closer to the magic value Φ).

I'm happy to pull that bit of code out into a separate function.  It's only in `OPENSSL_sk_reserve` at the moment.  I wanted to avoid cluttering that function with a long comment about the rationale behind the ratio choice and the concerns this raises.

levitte replied:
Is the exact rounding a serious reason for worry here?

dot-asm replied:
> Is the exact rounding a serious reason for worry here?

No, but I'd say it's appropriate to explicitly discuss it, so that if "but-what-about" question is risen later, there is record of conscious discussion.

paulidale replied:
The most accurate downward rounding has been done.  The off by one should be gone.

paulidale commented:
The expansion size calculation code is separated into its own function.
I've implemented the x+x/2 suggestion of @dot-asm 
I've also fixed the range checking to avoid the off by one low situation that could sometimes occur with the older code.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -21,10 +28,43 @@ struct stack_st {
+        sz = n < limit ? n + n / 2 : INT_MAX;
+
+        /* Overflow check, which is possible if SIZE_MAX < INT_MAX */
+        if (sz < n)
```
Well, I see contradiction here. If you consider SIZE_MAX < INT_MAX as possibility, then you can't calculate limit the way you do. I mean there is implicit cast to narrower type, so which value would *that* yield? If we make such assumption, then limit should be two thirds of *minimum* between SIZE_MAX/sizeof(void *) and INT_MAX. Note that it incidentally works on ILP32 systems... But note that we don't have to make such assumption, or rather we can *choose* not to make one by declaring that ILP32 is the minimum supported. [Well, formally speaking one can still argue that size_t can be narrower than 32 bits, because there is no S in ILP.] Myself I'm fine  with *not* making the assumption.

paulidale replied:
Good catch.

Are you suggesting to set the limit to the minimum?
Or that the SIZE_MAX >= INT_MAX assumption is reasonable?  Like you, I'm not aware of a machine that doesn't have this.

How about defining a constant equal to the minimum of SIZE_MAX/sizeof(char *) and INT_MAX and using this instead of both SIZE_MAX and INT_MAX for the limit checking?

dot-asm replied:
> Are you suggesting to set the limit to the minimum?
> Or that the SIZE_MAX >= INT_MAX assumption is reasonable?

Well, what I'm saying is that *IF* we make the assumption that SIZE_MAX can be less than INT_MAX, then we wouldn't have another choice. Then I'm suggesting *not* to make the assumption, because it's reasonable in the landscape of supported configurations(\*). Then I'm saying that minimum(\*\*) might be sensible *anyway* for ILP32 account. In other words I kind of suggest both: set to minimum(\*\*), but without making the assumption. I.e. it's not "or", but "and".

BTW, not making the assumption means that this `if (sz < n)` would have to be omitted. Basically making or not making the assumption is also question of readability. As discussed earlier, readability, among other things, is about how many questions pops in your mind when you read the code, and less is better. What I'm hinting at is that more assumptions lead to less readable code and this `if` is the illustration. I mean when you see it you have to ask yourself "why" and seek for answer [which would be "not applicable in this case" for vast majority].

(\*) Not to mention that it's likely to be made all over anyway. It might be appropriate to reflect the fact that we expect SIZE_MAX to be not less than INT_MAX in test/sanitytest.c, so that we would know when it breaks. [And if it does, we would still have an option to disclaim support, as opposite to fixing the whole thing.]

(\*\*) minimum between SIZE_MAX/sizeof(void *) and INT_MAX, not SIZE_MAX and INT_MAX.

dot-asm replied:
> set to minimum, but without making the assumption

Just in case, reason for that is kind of readability. We have to check for SIZE_MAX/sizeof(void*) in either case, don't we? And I'd say that it would be more appropriate if all checks are performed in one place, instead of two. I mean now compute_growth checks for INT_MAX, while SIZE_MAX/sizeof(void*) boundary is enforced elsewhere...

paulidale replied:
Using a single limit value simplified the range checking quite a bit.
I've moved the limit calculation code down next to the reserve function so the range checks are in one place.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -107,43 +147,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
+        || (size_t)n > SIZE_MAX / sizeof(char *) - st->num)
+        return 0;
+
+    if ((size_t)st->num + n >= st->num_alloc) {
```
Even here *if* one make assumption that size_t is narrower than int, then you'd be better of with implicit casts.

dot-asm replied:
Well, I bet warning system will complain about signed/unsigned comparison, so implicit cast won't do, right? But in such case it would be formally correct to cast left part to unsigned int and let implicit casts play out between unsigned int and size_t.

dot-asm replied:
Again, *if* we make such assumption. Though one still argue that unsigned int cast of the left part is more readable.

paulidale replied:
It was a build warning that prompted me to add the cast.
unsigned int seems reasonable and it is guaranteed to work.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -21,10 +28,43 @@ struct stack_st {
+    const size_t limit = (INT_MAX / 3) * 2 + INT_MAX % 3;
+    size_t sz, n = current;
+
+    while (n <= target) {
```
BTW, why `<=` and not `<`? Whole idea behind reservation is that you're telling that you're pretty sure that `target` is all you'll ever need. Then why would you insist on 50% overhang?

dot-asm replied:
With this in mind one can actually wonder if reservation should even go through this loop. I mean if you're pretty sure that `target` is all you'll *ever* need, then why have *any* overhang? Even if it turns out that it needs extension after all [as result of single insertion], one can increment from the point of initial reservation.

paulidale replied:
I'll see if I can figure something out so the space isn't wasted.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -21,10 +28,43 @@ struct stack_st {
+
+    while (n <= target) {
+        /* This shouldn't be possible but check it anyway */
+        if (n >= INT_MAX)
```
The commentary is misleading. Let's say `target` was INT_MAX. It should be legitimate value on 64-bit system [just legitimate, not practical]. But note that given `<=` loop condition and sz assignment below, it will be rejected. Once again, this is not about what's practical, but what's *formally* correct.

paulidale replied:
This one shouldn't be possible because of the earlier checks that detect this case.
The loop can set `n` as high as INT_MAX and if that's insufficient the request would have been denied earlier.  I don't like the possibility of infinite loops however.

I'll improve the comment.

dot-asm replied:
In the loop body you have sz = ... INT_MAX, n = sz, then go to loop beginning, which is while (n <= target). If target is INT_MAX, it enters the loop and returns 0. When it should have returned INT_MAX.

> would have been denied earlier.

Earlier > INT_MAX is denied, not >= INT_MAX.

paulidale replied:
I'll fix the `<=` check.  It should be '<'.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
-        return 0;
+    const int limit = (max_nodes / 3) * 2 + max_nodes % 3;
```
Now when max_nodes is "variable" `max_nodes % 3` term appears somewhat controversial. For known values of 0x7fffffff and 0xffffffff/4 it works, but not in general case. "Works" means that `limit + limit/2` doesn't actually exceed the max_nodes. Or in other words *in general case* you can find a max_nodes value for which `limit + limit/2` would exceed max_nodes. So that preceding commentary is actually misleading, because it gives an impression of general case. On the other hand we are not performing `limit + limit/2` operation, but `limit-1 + (limit-1)/2`... So I suppose suggestion is to either omit the term, or clarify commentary by stating that we are scaling up limit-1 [because it's `current < limit` and not `current <= limit`].

paulidale replied:
Ouch!  2^n mod 3 is always 0 or 1.  The code as written overestimates by 1 when `max_nodes` mod 3 = 2.  Good catch.

dot-asm replied:
> 2^n mod 3 is always 0 or 1

??? 2^3 % 3 is 2.

paulidale replied:
2^n-1 was the intention.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
+
+    while (current <= target) {
+        /* Check to see if we're at the hard limit */
+        if (current >= max_nodes)
```
Once again, in combination with loop condition this prohibits *formally* legitimate case of target == max_nodes.

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
+    num_alloc = st->num + n;
+    if (!exact) {
+        if (num_alloc <= st->num_alloc)
+            return 1;
```
Speaking of this condition. Wouldn't it be appropriate to check for equality in exact case? [For reference, my first reaction was to suggest to take out of `if (!exact)`, but then I realized that it would prevent freeing some memory, which might be desirable.]

paulidale replied:
Does the exact + equality fall out with the later `num_alloc == st->num` test? a few lines below?

dot-asm replied:
Oh! But no, it doesn't cut it. Because it's different condition, and one can even wonder what's up with that. What does size of array have to do number of initialized elements? I mean I'm struggling with question why look at st->num at all. At least at reservation time it's likely to be 0, isn't it?

dot-asm replied:
It the context one can wonder if it makes sense to postpone allocation of st->data till first reservation or insertion. On can also wonder if it's appropriate to increase min_nodes. Every allocation has an overhead and question is how does min_modes of 4 compare to this overhead. I mean it's desirable that overhead/payload ratio is low.

paulidale replied:
I see the problem now.  Is there such a thing as _code blindness_?
I've fixed this and moved the minimum test up a bit.

As for postponing the allocation of st->data, it is probably a good idea.
Likewise, `min_nodes` of 4 seems low to be but without some statistics it is impossible to tell.
I'd suggest a separate PR for each.

paulidale replied:
I've created #4426 to track the suggestions for the future.

dot-asm commented:
While you're at it, you can fix formatting in OPENSSL_sk_deep_copy declaration...

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
+
+int OPENSSL_sk_reserve(OPENSSL_STACK *st, int n)
+{
+    if (st == NULL || st->num < 0)
```
I'd like to argue in favour of omission of st == NULL check. Here and elsewhere. Thoughts?

paulidale replied:
I'm unsure.  Both the `st == NULL` and `st->num < 0` seem redundant which an opaque structure.
This PR isn't really about cleaning the code up, so I left them unchanged.  I don't want to clutter reviews least something is missed in the range code.

There are quite a number of `return (x)` instances in the file which I've also left alone.

dot-asm replied:
Not doing it here and now is fine. Omitting even st->num < 0... Hmmm... Yeah, I suppose one can omit even that...

dot-asm commented on crypto/stack/stack.c:
```diff
@@ -107,43 +112,98 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
+        if (num_alloc == 0)
             return 0;
+    } else if (num_alloc == st->num_alloc)
+        return 1;
```
Style nit. If `if` has {}, then `else` has to have pair too.

[Off for today...]

paulidale replied:
Fixed.

paulidale commented:
Thanks for persevering on this one and find niggling off by one problems.
Merged.

richsalz commented:
Perhaps take a look at and close #4379 ?
